dask/order.py

```diff
@@ -202,125 +193,212 @@ def dependencies_key(x):
-
-            while not outer_stack and outer_stack_seeds_index < len(outer_stack_seeds):
-                # We wait for as long as possible to consider dependents of completed nodes:
-                # 1. some may have already completed, so waiting lets us sort fewer items, and
-                # 2. sorting via `dependents_key` depends on `num_needed`, which is dynamic.
-                deps = set_difference(
-                    dependents[outer_stack_seeds[outer_stack_seeds_index]], result
-                )
-                if deps:
-                    if 1 < len(deps) < 1000:
-                        outer_stack_extend(
-                            sorted(deps, key=dependents_key, reverse=True)
-                        )
-                    else:
-                        outer_stack_extend(deps)
-                outer_stack_seeds_index += 1
+                            min_key = min(dep_pools)
+                            min_pool = dep_pools[min_key]
+                            if 1 < len(min_pool) < 100:
+                                dep = min(min_pool, key=dependents_key)
+                            else:
+                                dep = min_pool.pop()
+                                if not min_pool:
+                                    del dep_pools[min_key]
+                            inner_stack = [dep]
+                            for key, vals in dep_pools.items():
+                                if key < item_key:
+                                    if key in next_nodes:
+                                        next_nodes[key].append(vals)
+                                    else:
+                                        next_nodes[key] = [vals]
+                                elif key in later_nodes:
+                                    later_nodes[key].append(vals)
+                                else:
+                                    later_nodes[key] = [vals]
 
         if len(dependencies) == len(result):
             break  # all done!
 
+        if next_nodes:
+            if len(next_nodes) > 150:
+                # convert to log scale
+                new_next_nodes = {}
+                for key, vals in next_nodes.items():
+                    key = int(log(key + 1))
+                    if key in new_next_nodes:
+                        new_next_nodes[key].extend(vals)
+                    else:
+                        new_next_nodes[key] = vals
+                next_nodes = new_next_nodes
+
+                while len(next_nodes) > 150:
+                    denom = len(next_nodes) // 75
+                    new_next_nodes = {}
+                    for key, vals in next_nodes.items():
+                        key //= denom
+                        if key in new_next_nodes:
+                            new_next_nodes[key].extend(vals)
+                        else:
+                            new_next_nodes[key] = vals
+                    next_nodes = new_next_nodes
```

Comment:
This block appears to be untested.

Reply:
Does it count if lines are covered by benchmarks?  But, yes, it would be best to have more full coverage directly from tests.

Reply:
It's typically good to have test coverage (if possible). The lag time between a commit introducing a performance regression and someone actually noticing it can be large :)

Reply:
The first block here (the one that takes the logarithm) gets run via `dask/array/tests/test_linalg.py`.  The second block is still untested.